tool/synthetic-data-sampler.ts

```diff
@@ -0,0 +1,530 @@
+            canonicalAnnotation.adjective = ["# 's", '#'];
+
+        // if it's the only date, adding argmin/argmax/base_projection
+        if (typestr === 'Date') {
+            canonicalAnnotation.adjective_argmax = ["most recent", "latest", "last", "newest"];
+            canonicalAnnotation.adjective_argmin = ["earliest", "first", "oldest"];
+            canonicalAnnotation.base_projection = ['date'];
+        }
+    }
+    return canonicalAnnotation;
+}
+
+function generateFilterAst(device : string, func : string, property : string, operator : string, value : any) : Ast.Program {
+    const invocation = new Ast.InvocationExpression(
+        null, 
+        new Ast.Invocation(null, new Ast.DeviceSelector(null, device, null, null), func, [], null),
+        null
+    );
+    const filter = new Ast.AtomBooleanExpression(
+        null,
+        property, 
+        operator, 
+        value, // 
+        null
+    );
+    const filtered = new Ast.FilterExpression(
+        null, 
+        invocation, 
+        filter, 
+        null
+    );
+    const statement = new Ast.ExpressionStatement(
+        null,
+        new Ast.ChainExpression(null, [filtered], null)
+    );
+    return new Ast.Program(
+        null,
+        [], 
+        [],
+        [statement],
+        {}
+    );
+}
+
+
+function generateProjectionAst(device : string, func : string, property : string) : Ast.Program {
+    const invocation = new Ast.InvocationExpression(
+        null, 
+        new Ast.Invocation(null, new Ast.DeviceSelector(null, device, null, null), func, [], null),
+        null
+    );
+    const projection = new Ast.ProjectionExpression(
+        null,
+        invocation, 
+        [property],
+        [], 
+        [], 
+        null
+    );
+    const statement = new Ast.ExpressionStatement(
+        null,
+        new Ast.ChainExpression(null, [projection], null)
+    );
+    return new Ast.Program(
+        null,
+        [], 
+        [],
+        [statement],
+        {}
+    );
+}
+
+function toTSV(device : string, data : NewParaphraseExample[], useHeading : boolean) {
+    let headings : string = '';
+    if (useHeading)
+        headings = ["id", "utterance", "thingtalk", "query", "queryCanonical", "argument", "value"].join('\t') + '\n';
+    const rows = data.reduce((acc : string[], colValue, idx) => {
+        const id = `${device}-${idx.toLocaleString('en-US', {minimumIntegerDigits: 3, useGrouping:false})}`;
+        const tmp = [
+            id,
+            colValue["utterance"],
+            colValue["thingtalk"]
+        ]
+        return acc.concat([tmp.join('\t')]);
+    }, []).join('\n');
+    return `${headings}${rows}`;
+}
+
+function toThingtalkValue(value : string, type : Type) : { value : Ast.Value, op : string } {
+    value = value.toLowerCase();
+    if (type instanceof Type.Entity) {
+        value = `"${value}"`;
+        return {value: new Ast.Value.Entity(value, 'tt:device', value), op: "=~" };
```

Comment:
`display` in general is not the same as `value`, the constant should come with both value and display for entity type, and here should load that properly. 
entity filters should always use `==`. `=~` is only used when the constat is a string